New: Add "exceptions" option to space-infix-ops

evnp opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

AST node types to be excluded from space-infix-ops can now be
defined via "exceptions". Follows precedent and conventions set by
no-multi-spaces. Use case: es6 default args a la "(a=0) => {}"

**Is there anything you'd like reviewers to focus on?**

Nope

evnp commented:
Ah, completely forgot that. Thanks @ilyavolodin, updated the docs

not-an-aardvark commented:
Related: https://github.com/eslint/eslint/pull/7591

evnp commented:
Realized I used the wrong PR template initially, I've updated. Anything I can do to move this along? FWIW I don't think https://github.com/eslint/eslint/pull/7591 handles my use case as I'd like to allow `(a=0) => {}` while still disallowing `let a=0;`

ilyavolodin commented:
@eslint/eslint-team Does anyone wants to support this?

platinumazure commented:
I'm a bit confused about how exceptions actually work. What happens when you use false as the value? Is it possible to *forbid* spacing except for named node types? And how would a user know what node types are even valid (e.g., if a user specifies Identifier or Literal, it won't be much use)?

Until those questions are answered, I can't really get behind this.

evnp commented:
They work in the same way as the "exceptions" option for no-multi-spaces http://eslint.org/docs/rules/no-multi-spaces. Setting a node type to `false` would cause it to not be treated as an exception to the rule, so spaces would be enforced (as is the case for no-multi-spaces). Copying no-multi-spaces, the docs suggest using the [online demo](http://eslint.org/parser/) and experimentation to determine what node types are valid for the rule (it won't apply to node types the rule doesn't apply to, but I believe this is also the case with no-multi-spaces). 

I hope that clears things up. This seems like a useful change since it solves a currently-unenforcable style requirement of mine (python-style default arg spacing) which I don't think is covered by other proposals, and the precedent already seems to be set by no-multi-spaces.

evnp commented:
btw here's the impetus for this, thought it might be helpful https://github.com/mixpanel/eslint-config/pull/8. At the end of the day I just want to be able to support `(a=0) => {}` and `let a=0;` at the same time, happy to change the interface here if it makes more sense. "exceptions" just seems nicely consistent with no-multi-spaces

ilyavolodin commented:
Another ping to @eslint/eslint-team anyone wants to support this option? If not, we should probably close this.

not-an-aardvark commented:
I think having an option for this is a good idea in general, but we should make sure the schema works well with the changes in https://github.com/eslint/eslint/pull/7591. Instead of exceptions, maybe we should allow things like `AssignmentPattern` to be configured independently, e.g. `AssignmentPattern: never`.

evnp commented:
@not-an-aardvark @ilyavolodin sorry I dropped the ball on this. Agree that having finer-grained control is much better, since you can actually enforce lack of spacing in the AssignmentPattern `foo(bar=baz)` case I'm concerned about. I've added some changes to make this possible with "always", "never", "ignore" exception values, please let me know what you think.

kaicataldo commented:
Closing this since there hasn't been any discussion in quite a while and progress on the issue has stalled. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. 

Thank you for contributing to ESLint, regardless!
